shared/src/main/scala/mlscript/Typer.scala

```diff
@@ -365,201 +376,239 @@ class Typer(var dbg: Boolean, var verbose: Bool, var explainErrors: Bool) extend
           implicit val prov: TypeProvenance = mt.prov
           println(s">> Checking subsumption for inferred type of $nme : $mt")
-          ctx.mthDecls.get(tn.name).foreach(_.get(nme).foreach(ss(mt, _)))
+          (if (decls.isDefinedAt(nme) && !defns.isDefinedAt(nme)) decls.get(nme) else N).orElse(declsInherit.get(nme))
+            .foreach(ss(mt, _))
+          val mthTy = td.wrapMethod(mt)
+          if (!declsInherit.get(nme).exists(decl => decl.single && decl.parents.last === mt.parents.last))
+            ctx.addMth(S(tn.name), nme, mthTy)
         }
-        (td.mthDecls ++ td.mthDefs).foreach { case md @ MethodDef(rec, prt, nme, tparams, rhs) =>
-          implicit val prov: TypeProvenance = tp(md.toLoc, rhs.fold(_ => "method definition", _ => "method declaration"))
-          val mt = rhs.fold(_ => defs(nme.name), _ => decls(nme.name))
-          td.baseClasses.foreach { case Var(bn) => 
-            ctx.mthDefs.get(bn).foreach(_.get(nme.name).foreach { bmt =>
-              if (!ctx.mthDecls.get(bn).exists(_.isDefinedAt(nme.name))) {
-                err(msg"Overriding method ${bn}.${nme.name} without explicit declaration is not allowed.", md.toLoc)
-                println(s">> Checking subsumption (against inferred type) for inferred type of $nme : $mt")
-                ss(mt, bmt)
-              }
-            })
+        decls.foreach { case nme -> mt =>
+          implicit val prov: TypeProvenance = mt.prov
+          println(s">> Checking subsumption for declared type of $nme : $mt")
+          ((declsInherit.get(nme), defnsInherit.get(nme)) match {
+            case (S(decl), S(defn)) =>
+              if (mt.body =/= decl.body && !defns.isDefinedAt(nme)) defn.parents.foreach(parent => 
+                err(msg"Overriding method ${parent}.${nme} without an overriding definition is not allowed." -> mt.prov.loco ::
+                  msg"Note: method definition inherited from" -> defn.prov.loco :: Nil)(raise, noProv))
+              S(decl)
+            case (S(decl), N) => S(decl)
+            case (N, S(defn)) =>
+              overrideError(nme, mt, defn)
+              S(defn)
+            case (N, N) => N
+          }).foreach(ss(mt, _))
+          val mthTy = td.wrapMethod(mt)
+          ctx.addMth(S(tn.name), nme, mthTy)
+          registerImplicit(nme, mthTy)
+        }
+        defns.foreach { case nme -> mt => 
+          implicit val prov: TypeProvenance = mt.prov
+          println(s">> Checking subsumption for inferred type of $nme : $mt")
+          decls.get(nme).orElse((declsInherit.get(nme), defnsInherit.get(nme)) match {
+            case (S(decl), S(defn)) if defn.parents.toSet.subsetOf(decl.parents.toSet) => S(decl)
+            case (_, S(defn)) =>
+              overrideError(nme, mt, defn)
+              S(defn)
+            case (S(decl), N) => S(decl)
+            case (N, N) => N
+          }).foreach(ss(mt, _))
+          val mthTy = td.wrapMethod(mt)
+          if (!decls.isDefinedAt(nme)) {
```

Comment:
```suggestion
          if (!decls.isDefinedAt(nme)) {
            // If the class does not declare that method explicitly (it is inherited),
            // we still want to make the method available
            // both for implicit method calls (of the form `x.Foo`)
            // and for explicit ones (`x.(TheClass.Foo)`)
```

Reply:
Addressed in c5f6b0817ec9905325a697a9281652782b2d2e30. The examples for explicit vs. implicit calls are removed as they were documented elsewhere.